gtsfm/common/gtsfm_data.py

```diff
@@ -0,0 +1,196 @@
+
+        Args:
+            track: track to add.
+
+        Returns:
+            Flag indicating the success of adding operation.
+        """
+        # check if all cameras are already added
+        for j in range(track.number_measurements()):
+            i, _ = track.measurement(j)
+
+            if i not in self._cameras:
+                return False
+
+        self._tracks.append(track)
+        return True
+
+    def add_camera(self, camera: PinholeCameraCal3Bundler, camera_index: int) -> None:
+        """Adds a camera.
+
+        Args:
+            camera: camera object to it.
+            camera_index: the index associated with this camera.
+
+        Raises:
+            ValueError: if the camera to be added is not a valid camera object.
+        """
+        if camera is None:
+            raise ValueError("Camera cannot be None, should be a valid camera")
+        self._cameras[camera_index] = camera
+
+    def drop_cameras(self, camera_indices: List[int]) -> "GtsfmData":
+        """Drops cameras and tracks associated with them.
+
+        Returns:
+            New object with the requested cameras (and associated tracks) dropped.
+        """
+
+        new_data = GtsfmData(self._number_images)
+
+        for i, camera in self._cameras.items():
+            if i not in camera_indices:
+                new_data.add_camera(camera, i)
+
+        new_camera_indices = new_data.get_valid_camera_indices()
+
+        # add tracks which have all the camera present in new data
+        for track in self._tracks:
+            is_valid = True
+            for m_idx in range(track.number_measurements()):
+                i, _ = track.measurement(m_idx)
+                if i not in new_camera_indices:
+                    is_valid = False
+                    break
+            if is_valid:
+                new_data.add_track(track)
+
+        return new_data
+
+    def select_largest_connected_component(self) -> "GtsfmData":
+        """Selects the subset of data belonging to the largest connected component of the graph where the edges are
+        between cameras which feature in the same track.
+
+        Returns:
+            New GtSfmData object with the subset of tracks and cameras.
+        """
+        camera_edges = []
+        for sfm_track in self._tracks:
+            cameras_in_use = []
+            for m_idx in range(sfm_track.number_measurements()):
+                i, _ = sfm_track.measurement(m_idx)
+                cameras_in_use.append(i)
+
+            # add all combinations of edges
+            camera_edges += list(itertools.combinations(cameras_in_use, 2))
+
+        if len(camera_edges) == 0:
+            return GtsfmData(self._number_images)
+
+        cameras_in_largest_cc = graph_utils.get_nodes_in_largest_connected_component(camera_edges)
+
+        # select cameras to drop and drop them
+        dropped_cameras = np.setdiff1d(list(self._cameras.keys()), cameras_in_largest_cc)
```

Comment:
@ayushbaid is it simpler to make the constructor with the "active cameras", instead of the "dropped ones"? @akshay-krishnan does that seem cleaner to you?